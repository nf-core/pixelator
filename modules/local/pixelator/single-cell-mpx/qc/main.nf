process PIXELATOR_QC {
    tag "${meta.id}"
    label 'process_medium'

    // TODO: Add conda back
    // conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'quay.io/pixelgen-technologies/pixelator:0.21.2'
        : 'quay.io/pixelgen-technologies/pixelator:0.21.2'}"

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path("adapterqc/*.processed.{fq,fastq}.gz"),      emit: processed

    tuple val(meta), path("adapterqc/*.processed.{fq,fastq}.gz"),      emit: adapterqc_processed
    tuple val(meta), path("preqc/*.processed.{fq,fastq}.gz"),          emit: preqc_processed

    tuple val(meta), path("adapterqc/*.failed.{fq,fastq}.gz"),         emit: adapterqc_failed
    tuple val(meta), path("preqc/*.failed.{fq,fastq}.gz"),             emit: preqc_failed
    tuple val(meta), path("{adapterqc,preqc}/*.failed.{fq,fastq}.gz"), emit: failed

    tuple val(meta), path("adapterqc/*.report.json"),                  emit: adapterqc_report_json
    tuple val(meta), path("preqc/*.report.json"),                      emit: preqc_report_json
    tuple val(meta), path("{adapterqc,preqc}/*.report.json"),          emit: report_json

    tuple val(meta), path("preqc/*.qc-report.html"),                   emit: preqc_report_html

    tuple val(meta), path("adapterqc/*.meta.json"),                    emit: adapterqc_metadata
    tuple val(meta), path("preqc/*.meta.json"),                        emit: preqc_metadata
    tuple val(meta), path("{adapterqc,preqc}/*.meta.json"),            emit: metadata

    tuple val(meta), path("*pixelator-preqc.log"),                     emit: preqc_log
    tuple val(meta), path("*pixelator-adapterqc.log"),                 emit: adapterqc_log
    tuple val(meta), path("*pixelator-*.log"),                         emit: log

    path "versions.yml",                                               emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    assert meta.design

    def prefix = task.ext.prefix ?: "${meta.id}"
    def preqc_args = task.ext.args ?: ''
    def adapterqc_args = task.ext.args2 ?: ''

    // --design is passed in meta and added to args and args2 through modules.conf
    """
    pixelator \\
        --cores ${task.cpus} \\
        --log-file ${prefix}.pixelator-preqc.log \\
        --verbose \\
        single-cell-mpx \\
        preqc \\
        --output . \\
        ${preqc_args} \\
        ${reads}

    shopt -s nullglob
    preqc_results=( preqc/*.processed.* )
    echo \${preqc_results[@]}
    shopt -u nullglob # Turn off nullglob to make sure it doesn't interfere with anything later

    pixelator \\
        --cores ${task.cpus} \\
        --log-file ${prefix}.pixelator-adapterqc.log \\
        --verbose \\
        single-cell-mpx \\
        adapterqc \\
        --output . \\
        ${adapterqc_args} \\
        \${preqc_results[@]}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mkdir preqc
    echo "" | gzip >> "preqc/${prefix}.processed.fq.gz"
    echo "" | gzip >> "preqc/${prefix}.failed.fq.gz"
    touch "preqc/${prefix}.report.json"
    touch "preqc/${prefix}.meta.json"
    touch "preqc/${prefix}.qc-report.html"
    touch "${prefix}.pixelator-preqc.log"

    mkdir adapterqc
    echo "" | gzip >> "adapterqc/${prefix}.processed.fq.gz"
    echo "" | gzip >> "adapterqc/${prefix}.failed.fq.gz"
    touch "adapterqc/${prefix}.report.json"
    touch "adapterqc/${prefix}.meta.json"
    touch "${prefix}.pixelator-adapterqc.log"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """
}
