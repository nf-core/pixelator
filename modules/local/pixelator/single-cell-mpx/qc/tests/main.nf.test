nextflow_process {

    name "Test Process PIXELATOR_QC"
    script "../main.nf"
    process "PIXELATOR_QC"
    tag "modules"
    tag "pixelator"
    tag "pixelator/mpx"
    tag "pixelator/single_cell_mpx_qc"


    test("Test MPX qc - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'sample01_1k_pbmcs_scsp_v1_immunology1', design:'D21', panel:'human-sc-immunology-spatial-proteomics-1', technology:'mpx' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/amplicon/sample01_1k_pbmcs_scsp_v1_immunology1.merged.fastq.gz', checkIfExists: true),
                    ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Test MPX qc - SCSP v1 | Immunology-I") {
        when {
                params {
                save_all = true
            }
            process {
                """
                input[0] = [
                    [ id:'sample01_1k_pbmcs_scsp_v1_immunology1', design:'D21', panel:'human-sc-immunology-spatial-proteomics-1', technology:'mpx' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/amplicon/sample01_1k_pbmcs_scsp_v1_immunology1.merged.fastq.gz', checkIfExists: true),
                    ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                        process.out.adapterqc_processed,
                        process.out.preqc_processed,
                        process.out.adapterqc_failed,
                        process.out.preqc_failed,
                        process.out.adapterqc_report_json,
                        process.out.preqc_report_json,
                        process.out.adapterqc_metadata,
                        process.out.preqc_metadata,
                        process.out.versions
                    )
                },
                {
                    with(process.out.preqc_report_html) {
                        assert path(get(0).get(1)).exists()
                    }
                },
                {
                    with (process.out.preqc_log) {
                        assert path(get(0).get(1)).readLines().any { it.contains("Finished pixelator preqc") }
                    }
                },
                {
                    with (process.out.adapterqc_log) {
                        assert path(get(0).get(1)).readLines().any { it.contains("Finished pixelator adapterqc") }
                    }
                },
            )
        }
    }
}

