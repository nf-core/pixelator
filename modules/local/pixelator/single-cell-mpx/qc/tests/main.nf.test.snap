{
    "Test MPX qc - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.processed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.processed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "10": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "13": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-preqc.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "14": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-adapterqc.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "15": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-adapterqc.log:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-preqc.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "16": [
                    "versions.yml:md5,bdb6b9ae999e232953034e12b3835197"
                ],
                "2": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.processed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.qc-report.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "adapterqc_failed": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "adapterqc_log": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-adapterqc.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "adapterqc_metadata": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "adapterqc_processed": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.processed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "adapterqc_report_json": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "failed": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-adapterqc.log:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-preqc.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "metadata": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "preqc_failed": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.failed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "preqc_log": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-preqc.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "preqc_metadata": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "preqc_processed": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.processed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "preqc_report_html": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.qc-report.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "preqc_report_json": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "processed": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.processed.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "report_json": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        [
                            "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bdb6b9ae999e232953034e12b3835197"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-26T11:53:22.782634095"
    }
}